### ISSUE TITLE
[Feature] Built-in dictionary + redesign

### ISSUE THREAD

**Issue by realgooseman** [translate-tools/linguist#397]
Hi @vitonsky and contributors,

As the title says, a built-in dictionary that gives word definition, synonyms,...
For the sources, I personally use TheFreeDictionary and Wiktionary for my definitions as they are the easiest to use and accurate. But I don't know if they can get implemented or if they have a free API.

I made quick mockups of the dictionary feature + redesigned the translation pop-up to give ideas. 

## **How would it work ?**

If one word is selected on a page = translation + dictionary pop-up shows up
If a few words/sentence selected = only translation pop-up shows up

## **Translation pop-up redesign :**

![Pop-up redesign](https://github.com/translate-tools/linguist/assets/64812183/da25690c-5455-40a8-85ac-37786cb62a52)

## **Single word translation + dictionary :** 

![Dictionary](https://github.com/translate-tools/linguist/assets/64812183/fe957897-3bf8-4043-a59f-c2eabe5e925f)

**Comment by Terkyz** [comment]
This is a million times more elegant and modern than the current design, it looks absolutely great!

**Comment by vitonsky** [comment]
Looks good, thanks for your efforts!

I think it is not a big deal to find API for dictionary, so i will implement the feature in near next release.
Let's fix some cases in UI before i start to code it.

# Mobile appearance
Let's think how popup will appears on mobile devices.

For now mobile version uses Modal Window instead of Popup. Layout on mobile version are vertical with buttons that fit all horizontal space.

![image](https://github.com/translate-tools/linguist/assets/86191922/2c6456ba-320a-4f1e-baec-2e0a794b6a5c)

# Block with original text are loosed

For now the popup have block with original text. Here you can speak text on original language, it is important feature, so we have to keep it. Could you include this feature in your design.

![image](https://github.com/translate-tools/linguist/assets/86191922/9c8b4216-beac-452f-ba7b-5aa29ddad57b)

# Languages direction layout

We have to consider that languages names may be long. On my screenshot below it takes 2 words, but for some locales languages may have 4 words. We have to consider this cases.

I like the compact view on your "Single word translation + dictionary" screenshot, but in case if user will choose another language in list, popup have to expand, so TTS and Dictionary buttons will be moved, this is a bad when UI jump out of user cursor.

We gotta come up with something, to popup will never expand width after appearance.

![image](https://github.com/translate-tools/linguist/assets/86191922/9be44b1b-9d53-4e79-accf-6c204fe80ecf)

**Comment by realgooseman** [comment]
Linguist on mobile only works via Firefox right ? Because Chromium-based mobile browsers don't support extensions I think. 

Anyway, for mobile just copy the V2 mockups below and expand the blocks horizontally + put the buttons at the bottom.

## **Translation pop-up V2**

![Pop-up redesign V2](https://github.com/translate-tools/linguist/assets/64812183/433f030e-ad5c-48f2-b335-0eb725c089ca)

## **Translation + Dictionary pop-up V2**

![Dictionary V2](https://github.com/translate-tools/linguist/assets/64812183/b6be908d-5f14-4d63-b494-d9a35b3dbdf8)

Let me know what you think, I'm gonna sleep and be back tomorrow.

**Comment by realgooseman** [comment]
I agree, placing TTS button on textblock above the "copy to clipboard button" will be clean too. We get rid of the word counter and place a small volume button there. Should we replace the TTS button at the bottom with something else or just keep Bookmark/history buttons only ?

For the TTS button on dictionary block, where will it be placed ? Above the language & external link buttons ?
And should we also add TTS button for single translated word block on Dictionary pop-up or not (next to clipboard button) ?

Remembering the user's pop-up state to normal/expanded with original text might be a solution for people who need to always see the language pairs, or for the ones who only want to see the translation.

**Comment by vitonsky** [comment]
I think it's looks fine. I prefer to not split a block with original word and round borders not too much, but generally your design looks cool.

Your design also allow to implement editing text in popup in next iterations.

I will ask people from issue #301 to review your design. When we will have vision how it should looks, i will add it to queue to implement. Thanks for your work!

**Comment by realgooseman** [comment]
You're welcome, I'm just making quick mockups of what I think looks clean. But of course, you and the contributors should do what's best for Linguist. Code and usability wise.

Also, keep a dark theme variant in mind when working on the refreshed design. A lot of people's eyes would appreciate it.

**Comment by Flashwalker** [comment]
I vote for both buttons, tts and copy

**Comment by KiKaraage** [comment]
Still need that copy button tbh........

**Comment by QUIPQUOPRO** [comment]
Was this implemented? If so, are the built-in dictionary files stored with the browser extension files?

**Comment by vitonsky** [comment]
@QUIPQUOPRO hey, no this feature is not implemented yet.

To implement dictionary we need some source of definitions for a words on every supported language.

By Linguist nature it have no sense to implement dictionary only for words in English.
We need to cover at least top 10-20 languages.

If you know resources to fetch, please share it here. I will label current issue as "help wanted"
